File: bench/core/bench-concurrency.ts
// Concurrency scaling (TS side), mirroring the Rust `bench_concurrency.rs`.
// N workers each post M messages to a DISTINCT group (one live subscriber
// draining per group, untimed); we report aggregate posts/sec + deliveries/sec
// across concurrency levels.
//
// Runtime-model caveat (see the Rust module docs): Node runs on a single
// thread, so these workers run their sync post loops sequentially regardless of
// `concurrency` — throughput is capped near one core. The Rust mirror uses the
// multi_thread runtime. So this comparison shows each impl's CONCURRENCY-SCALING
// ceiling, not an apples-to-apples runtime match. For writes the coordinator is
// single-writer by design (mutex), so even Rust serializes on the post path —
// the headline here is per-op efficiency, not parallelism. See
// docs/benchmark-results.md.
//
// Run:
//   NODE_PATH=<repo>/references/cordn/node_modules node bench/core/bench-concurrency.ts
// Env: CORDN_BENCH_CONCURRENCY ("1,4,16,64"), CORDN_BENCH_MESSAGES (500),
//      CORDN_BENCH_BACKEND (both | sqlite | memory; memory recommended).

import { Coordinator } from "../../references/cordn/src/coordinator/coordinator.ts";
import { InMemoryCoordinatorStorage } from "../../references/cordn/src/coordinator/storage/inMemoryStorage.ts";
import { SqliteCoordinatorStorage } from "../../references/cordn/src/coordinator/storage/sqliteStorage.ts";
import { createPrivateMessage } from "../../references/cordn/src/coordinator/testUtils.ts";

function readPositiveIntEnv(name, fallback) {
  const raw = process.env[name];
  if (!raw) {
    return fallback;
  }
  const parsed = Number.parseInt(raw, 10);
  if (!Number.isFinite(parsed) || parsed <= 0) {
    throw new Error(`Environment variable ${name} must be a positive integer`);
  }
  return parsed;
}

function readConcurrencyLevels() {
  const raw = process.env.CORDN_BENCH_CONCURRENCY;
  if (!raw) {
    return [1, 4, 16, 64];
  }
  return raw
    .split(",")
    .map((s) => Number.parseInt(s.trim(), 10))
    .filter((n) => Number.isFinite(n) && n > 0);
}

function readBackends() {
  switch (process.env.CORDN_BENCH_BACKEND) {
    case "sqlite":
      return ["sqlite"];
    case "memory":
      return ["memory"];
    default:
      return ["memory"]; // sqlite is single-writer; default to memory
  }
}

async function consumeRecords(messages, expectedCount) {
  const iterator = messages[Symbol.asyncIterator]();
  try {
    for (let index = 0; index < expectedCount; index += 1) {
      const result = await iterator.next();
      if (result.done) {
        throw new Error(`ended early after ${index}, expected ${expectedCount}`);
      }
    }
  } finally {
    await iterator.return?.();
  }
}

async function runLevel(backend, concurrency, messages) {
  const storage =
    backend === "memory"
      ? new InMemoryCoordinatorStorage()
      : new SqliteCoordinatorStorage({ path: ":memory:" });
  const coordinator = new Coordinator({ storage });

  const drainers = []; // untimed correctness checks (async, await iterator.next())
  const workerFns = []; // timed post loops (deferred so timing is honest)
  for (let w = 0; w < concurrency; w += 1) {
    const gid = `g${w}`;
    const subscription = coordinator.subscribeGroupMessages({ groupId: gid });
    drainers.push(consumeRecords(subscription.messages, messages));
    // Deferred (not IIFE'd): the post loop has no `await`, so on Node's single
    // thread these run sequentially regardless of `concurrency`. That is the
    // honest TS result — no write parallelism — which is the point of the
    // comparison with the multi_thread Rust mirror.
    workerFns.push(() => {
      for (let i = 0; i < messages; i += 1) {
        coordinator.postGroupMessage({
          groupId: gid,
          opaqueMessage: createPrivateMessage({
            groupId: gid,
            epoch: 1n,
            contentType: 1,
            bytes: [i % 251],
          }),
        });
      }
    });
  }

  const start = process.hrtime.bigint();
  for (const fn of workerFns) {
    fn();
  }
  const totalNs = Number(process.hrtime.bigint() - start);
  await Promise.all(drainers); // untimed: every delivery landed
  storage.close();

  const secs = totalNs / 1_000_000_000;
  const total = concurrency * messages;
  return { totalMs: totalNs / 1_000_000, postsPerSec: total / secs, deliveriesPerSec: total / secs };
}

async function main() {
  const messages = readPositiveIntEnv("CORDN_BENCH_MESSAGES", 500);
  const levels = readConcurrencyLevels();

  console.log("TypeScript cordn — concurrency scaling (single-threaded event loop)");
  console.log(`messages/worker=${messages} concurrency=${JSON.stringify(levels)}`);
  console.log("");

  for (const backend of readBackends()) {
    console.log(`── ${backend} ──────────────────────────────`);
    console.log(
      `  ${"workers".padStart(10)} ${"total_ms".padStart(12)} ${"posts/sec".padStart(14)} ${"deliv/sec".padStart(14)}`,
    );
    for (const c of levels) {
      const r = await runLevel(backend, c, messages);
      console.log(
        `  ${String(c).padStart(10)} ${r.totalMs.toFixed(2).padStart(12)} ${r.postsPerSec.toFixed(0).padStart(14)} ${r.deliveriesPerSec.toFixed(0).padStart(14)}`,
      );
    }
  }
}

main().catch((error) => {
  console.error(error);
  process.exitCode = 1;
});

File: crates/cordn-core/examples/bench_concurrency.rs
//! Concurrency scaling. N worker tasks each post M messages to a DISTINCT group
//! (one live subscriber draining per group), then we report aggregate
//! posts/sec + deliveries/sec. Unlike `bench_fanout` (a single sequential
//! poster), this measures how throughput scales as parallel load increases.
//!
//! Runtime model caveat (important): this uses the multi_thread tokio runtime
//! so N workers genuinely run in parallel across cores. The TS coordinator is
//! inherently single-threaded (Node event loop), so the TS mirror cannot
//! parallelize — its throughput plateaus near one core. The comparison therefore
//! shows each impl's CONCURRENCY-SCALING ceiling, NOT an apples-to-apples
//! runtime-model match. That asymmetry is the point: it is the dimension the
//! sequential benches cannot surface. See docs/benchmark-results.md.
//!
//! Run: cargo run --release -p cordn-core --example bench_concurrency
//! Env: CORDN_BENCH_CONCURRENCY ("1,4,16,64"), CORDN_BENCH_MESSAGES (500),
//!      CORDN_BENCH_BACKEND (both | sqlite | memory; memory recommended —
//!      sqlite is single-writer so it serializes regardless of concurrency).

use std::env;
use std::sync::Arc;
use std::time::Instant;

use tokio::sync::Barrier;

use cordn_core::{
    Coordinator, CoordinatorOptions, CoordinatorStorage, InMemoryCoordinatorStorage,
    PostGroupMessageInput, SqliteCoordinatorStorage,
};

#[derive(Clone, Copy, PartialEq, Eq)]
enum Backend {
    Sqlite,
    Memory,
}

impl Backend {
    fn label(self) -> &'static str {
        match self {
            Backend::Sqlite => "SQLite",
            Backend::Memory => "In-memory",
        }
    }

    fn build(self) -> Arc<dyn CoordinatorStorage> {
        match self {
            Backend::Sqlite => Arc::new(
                SqliteCoordinatorStorage::open_in_memory()
                    .expect("open in-memory sqlite for bench"),
            ),
            Backend::Memory => Arc::new(InMemoryCoordinatorStorage::new()),
        }
    }
}

fn read_backends() -> Vec<Backend> {
    match env::var("CORDN_BENCH_BACKEND").ok().as_deref() {
        Some("sqlite") => vec![Backend::Sqlite],
        Some("memory") => vec![Backend::Memory],
        _ => vec![Backend::Memory], // sqlite is single-writer; default to memory
    }
}

fn read_pos_env(name: &str, fallback: usize) -> usize {
    match env::var(name) {
        Ok(raw) => raw
            .trim()
            .parse::<usize>()
            .ok()
            .filter(|v| *v > 0)
            .unwrap_or_else(|| panic!("{name} must be a positive integer, got {raw:?}")),
        Err(_) => fallback,
    }
}

fn read_concurrency_levels() -> Vec<usize> {
    match env::var("CORDN_BENCH_CONCURRENCY") {
        Ok(raw) => raw
            .split(',')
            .filter_map(|s| s.trim().parse::<usize>().ok().filter(|v| *v > 0))
            .collect(),
        Err(_) => vec![1, 4, 16, 64],
    }
}

// ponytail: fixed-size synthetic payload; revisit if results look size-sensitive.
fn opaque_message(index: usize) -> Vec<u8> {
    let mut bytes = vec![0u8; 32];
    bytes[31] = (index % 251) as u8;
    bytes
}

/// One concurrency level: N workers each post M messages to their own group;
/// one subscriber per group drains concurrently (untimed). Returns
/// (elapsed_ms, posts_per_sec, deliveries_per_sec).
async fn run_level(backend: Backend, concurrency: usize, messages: usize) -> (f64, f64, f64) {
    let coord = Coordinator::new(CoordinatorOptions {
        storage: Some(backend.build()),
        cleanup_interval_ms: Some(0),
        ..CoordinatorOptions::default()
    });

    // Barrier sized workers + main so timing starts exactly when all workers
    // are ready (no spawn-scheduling skew in the measurement).
    let barrier = Arc::new(Barrier::new(concurrency + 1));

    let mut drainer_handles = Vec::new();
    let mut worker_handles = Vec::new();
    for w in 0..concurrency {
        let gid = format!("g{w}");
        let mut sub = coord.subscribe_group_messages(&gid);

        // Drainer: untimed, just correctness (all messages received).
        drainer_handles.push(tokio::spawn(async move {
            for _ in 0..messages {
                sub.recv().await.expect("drain message");
            }
        }));

        // Worker: gated on the barrier, then posts its M messages.
        let coord_c = coord.clone();
        let barrier_c = barrier.clone();
        let gid_c = gid;
        worker_handles.push(tokio::spawn(async move {
            barrier_c.wait().await;
            for i in 0..messages {
                coord_c
                    .post_group_message(PostGroupMessageInput {
                        group_id: gid_c.clone(),
                        opaque_message: opaque_message(i),
                    })
                    .expect("post");
            }
        }));
    }

    // Release all workers together, time only the post loop.
    barrier.wait().await;
    let start = Instant::now();
    for handle in worker_handles {
        handle.await.expect("worker task");
    }
    let elapsed = start.elapsed();

    // Untimed correctness check: every delivery landed.
    for handle in drainer_handles {
        handle.await.expect("drainer task");
    }
    coord.close();

    let secs = elapsed.as_secs_f64();
    let total = (concurrency * messages) as f64;
    (secs * 1000.0, total / secs, total / secs)
}

// multi_thread so N workers run in parallel across cores (see module docs).
#[tokio::main(flavor = "multi_thread")]
async fn main() {
    let messages = read_pos_env("CORDN_BENCH_MESSAGES", 500);
    let levels = read_concurrency_levels();
    let backends = read_backends();

    println!("Rust cordn-core — concurrency scaling (multi_thread runtime)");
    println!(
        "messages/worker={messages} concurrency={:?} backend(s)={:?}",
        levels,
        backends.iter().map(|b| b.label()).collect::<Vec<_>>()
    );
    println!();

    for backend in backends {
        println!("── {} ──────────────────────────────", backend.label());
        println!(
            "  {:>10} {:>12} {:>14} {:>14}",
            "workers", "total_ms", "posts/sec", "deliv/sec"
        );
        for &c in &levels {
            let (ms, pps, dps) = run_level(backend, c, messages).await;
            println!("  {:>10} {:>12.2} {:>14.0} {:>14.0}", c, ms, pps, dps);
        }
    }
}
